### DISCUSSION TITLE
What is the equivalent of C# `delegate`?

### DISCUSSION THREAD

**Discussion by ghost** [pascalabcnet/pascalabcnet#3005] [Q&A] [answered]
I guess it's raw function pointers, but please correct me if I'm wrong.

**Comment by SunSerega** [comment, marked as answer]
It's a function pointer, but usually not raw.

You can just declare variables with their type being a `procedure`/`function`:
```
var a: function(x: byte): word;
var b: function(x: byte): word;
```
But note, while you can assign `a` to `b`, this will require a costly conversion, because `a` and `b` have their own delegate types.

So usually you instead declare a synonym for a delegate  type and then use that name:
```
type
  MyDelegate = function(x: byte): word;
  
begin
  var a: MyDelegate;
  var b: MyDelegate;
  //TODO a := ...
  b := a; // No conversion, just copy a reference
end.
```

There is also a base type `System.Delegate` that defines what all delegates have.

Meanwhile, a "raw" function pointer would be an actual pointer (or usually an `IntPtr`), pointed at the memory location where the executable processor commands are stored.
